lib/internal/modules/esm/resolve.js

```diff
@@ -726,6 +727,9 @@ function moduleResolve(specifier, base, conditions) {
  */
 function resolveAsCommonJS(specifier, parentURL) {
   try {
+    if (!CJSModule) {
```

Comment:
I had to lazy load this because there was a cycle with `lib/internal/modules/esm/resolve`

Reply:
It was causing `encodedSepRegEx` and `packageInternalResolve` to be undefined in [here](https://github.com/nodejs/node/blob/8618148073e311fcda75708fd7046e7de24e600a/lib/internal/modules/cjs/loader.js#L120). This, in turn, was breaking this [check](https://github.com/nodejs/node/blob/8618148073e311fcda75708fd7046e7de24e600a/lib/internal/modules/cjs/loader.js#L592) and this  [part](https://github.com/nodejs/node/blob/8618148073e311fcda75708fd7046e7de24e600a/lib/internal/modules/cjs/loader.js#L1055) . I did not test manually the last one, but looking at the code, it would have returned 'module not found' always if I am not mistaking.

Reply:
If there's a better way to tackle this, I am all ears.

Reply:
@dnlup I'm struggling to follow why this would be the case. There is a cycle that is carefully managed between these two modules yes - but it works because we only ever load the cjs/loader file first so that esm/resolve does not wait on cjs/loader execution and everything defines correctly.

The only reason why this cycle might bite is if you somehow cause esm/resolve to be loaded before cjs/loader.

But in this PR I see no changing of imports, so hence this seems very strange to me.

Reply:
@guybedford It seems to be happening only in `repl` mode. Could the order be different in that case?

Reply:
Yes that would make sense - all it would take is a `require` to esm/resolve to happen before a `require` to `cjs/loader` for this issue to happen. It might be possible to find out where that is?

Reply:
I'll try to find out, thanks for the pointers :wink:

Reply:
If you could just share the stack trace from the undefined case, that would help.

Reply:
Is this helpful?

![image](https://user-images.githubusercontent.com/15520377/87957401-7e492880-cab0-11ea-8b9d-bbebdce86760.png)

Reply:
I have increased the stack trace lines number:

![image](https://user-images.githubusercontent.com/15520377/87961100-89528780-cab5-11ea-9760-f1a5d84c98d8.png)

Reply:
Thanks that helps clarify the issue. It's caused by this line - https://github.com/nodejs/node/blob/master/lib/repl.js#L347, which is loading the ESM loader before the CJS loader and hence breaking the cycle references. This is a bug in core that will likely be possible to expose in other ways too.

I believe a nice way to fix this and try avoid it in future would be to include a `require('internal/modules/cjs/loader.js')` at the very top of https://github.com/nodejs/node/blob/master/lib/internal/modules/esm/loader.js#L1 to ensure that the cycles excute in the correct order, but my at-a-distance cycle logic might be incorrect too :)

Reply:
Thank you @guybedford , this looks like it fixes the cycle. I am wondering if there could be a way to refactor the two modules to avoid this but can't come up with any atm. If we are ok with this, I can resolve this discussion.